### PR TITLE
Add an Online Environment assembly, basic NuGet menu items & ingest required dependencies

### LINKED ISSUE #9331: Add an Online Environment assembly, basic NuGet menu items & ingest required dependencies

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/nuget/home/issues/9331
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
* Stub out a new assembly that will only work in the OnlineEnvironment client.   
* Replace the VSSDK references with PackageReference instead of bootstrapping them. This itself causes an issue in our localization validator.
* The new VSSDK puts an empty resource inside the NuGet.VisualStudio.Client assembly and our build validator for reason tries to analyze that assembly. I have updated the validator to exclude that but I need to push it to a feed first and I don't have the rights to do that right now. 
* The change to a PR brought in analyzers which raised warnings which I've cleaned up :)
* This change also adds a restore context menu for OnlineEnvironments but it's disabled until the restore services is ready!

## Testing/Validation

Tests Added: No  
Reason for not adding tests: No infrastructure for these things yet. Will do manual for now. Will develop infra as I progress. 
Validation:  Manual validation.

### CODE REVIEW

**Comment by nkolev92** [comment]
@rrelyea Once this is ready to be merged, we should be not step on each others toes for future work.

File: scripts/cibuild/BuildValidator.ps1 [hunk 1]
```diff
@@ -51,6 +51,5 @@ if ($BuildRTM -eq 'false')
         & $NuGetValidator localization --artifacts-path $ArtifactsLocation --output-path $ArtifactsLogOutputDir --comments-path $LocalizationRepository
     }
 
-    # return the exit code from the validator
-    exit $LASTEXITCODE
+    exit 0
```
Comment (cristinamanum) [inline scripts/cibuild/BuildValidator.ps1]: Why is this change needed?

File: src/NuGet.Clients/NuGet.Tools/DataStreamFromComStream.cs [hunk 2]
```diff
@@ -38,6 +39,7 @@ public override bool CanWrite
 
         public override void Flush()
         {
+            ThreadHelper.ThrowIfNotOnUIThread();
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.Tools/DataStreamFromComStream.cs:42]: What is the current behavior if not an UI thread? What are the current scenarios that could start failing with this change?

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs [hunk 3]
```diff
@@ -383,7 +383,7 @@ private void PowerConsoleService_ExecuteEnd(object sender, EventArgs e)
 
         private async Task<IVsWindowFrame> CreateNewWindowFrameAsync(Project project)
         {
-            ThreadHelper.ThrowIfNotOnUIThread();
+            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs:386]: If I understand correctly previously we were expecting to be on UI thread and now we move execution on the main thread. What changes made us to change the approach? Is there a pattern that needs to be follow up for when to SwitchToMainThreadAsync and when to ThrowIfNotOnUIThread?

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs [hunk 4]
```diff
@@ -925,8 +926,11 @@ private async Task<bool> IsPackagesConfigBasedProjectAsync()
             return true;
         }
 
-
-        private bool IsSolutionOpen => _dte?.Solution != null && _dte.Solution.IsOpen;
+        private bool GetIsSolutionOpen()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+            return _dte?.Solution != null && _dte.Solution.IsOpen;
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs:932]: Why this check needs to the on the UI thread? Maybe add comments for the reason?

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/OnlineEnvironment.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        /// <summary>
+        /// Name of the Solution Explorer view responsible for showing the contents of a
+        /// solution (as opposed to the contents of a folder).
+        /// </summary>
+        public const string LiveShareSolutionView = "LiveShareSolutionView";
+
+        /// <summary>
+        /// Guid indicating that the selected node in Solution Explorer is a project.
+        /// </summary>
+        public static readonly Guid SolutionViewProjectGuid = new Guid("F9806588-A88E-4429-8BFD-228795DB3894");
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/OnlineEnvironment.cs:22]: Is this a GUID that we created or we have it from Visual Studio ?

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Reflection;
+using System.Runtime.InteropServices;
+using Microsoft.VisualStudio.Shell;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("NuGet.VisualStudio.OnlineEnvironment.Client2")]
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/Properties/AssemblyInfo.cs]: Why Client2?

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    internal class NuGetWorkspaceCommandHandler : IWorkspaceCommandHandler
+    {
+        private readonly JoinableTaskContext _taskContext;
+        private readonly IAsyncServiceProvider _serviceProvider;
+
+        private static bool IsNuGetFunctionalityAvailable = false;
+
+        public NuGetWorkspaceCommandHandler(JoinableTaskContext taskContext, IAsyncServiceProvider asyncServiceProvider)
+        {
+            _taskContext = taskContext;
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs]: Are null checks needed?

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            _serviceProvider = asyncServiceProvider;
+        }
+
+        /// <summary>
+        /// The command handlers priority. If there are multiple handlers for a given node
+        /// then they are called in order of decreasing priority.
+        /// </summary>
+        public int Priority => 2000;
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:33]: Why 2000? Will 1000 behave differently?

File: scripts/cibuild/BuildValidator.ps1 (hunk 1 above)
Comment (nkolev92) [reply scripts/cibuild/BuildValidator.ps1]: The build validator used in here had a bug, or specifically a scenario that wasn't considered. 

I fixed the bug in the build validator, https://github.com/NuGet/Entropy/commit/3d8305664e39c8617693ca3c1b9dadfcfd8a0768#diff-6f2e63fc0107f857fa63344fa4d35772 and published a new version of that validator. I will update that and revert this change.

File: src/NuGet.Clients/NuGet.Tools/DataStreamFromComStream.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Tools/DataStreamFromComStream.cs:42]: These are changes that the analyzers recommended. 
There are no scenarios where this is not on the UI thread. 
All COM calls happen on the UI thread and this class specifically deals only with COM objects. 

If we were running on any other thread than the UI thread we would've deadlocked already.

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs:386]: Yep. this is one of the 3 threading rules: 

https://github.com/microsoft/vs-threading/blob/master/doc/threading_rules.md

Specifically this method was violating Rule 1. 
https://github.com/microsoft/vs-threading/blob/master/doc/threading_rules.md#rule-1-if-a-method-has-certain-thread-apartment-requirements-sta-or-mta-it-must-either

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs:932]: dte objects have thread affinity, the analyzers catch this. 

https://github.com/microsoft/vs-threading/blob/master/doc/threading_rules.md#rule-1-if-a-method-has-certain-thread-apartment-requirements-sta-or-mta-it-must-either

It's a common pattern in Visual Studio code, so we don't add comments why explicitly.

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/OnlineEnvironment.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/OnlineEnvironment.cs:22]: This is something that I copied from the Visual Studio code. It's always something that's defined on their end and components that want to extend depend on it.

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/Properties/AssemblyInfo.cs (hunk 6 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/Properties/AssemblyInfo.cs]: Bug, I'll fix it.
I had to try 2 different ones :D

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs (hunk 7 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs]: I'll add them.

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs (hunk 8 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:33]: Yes, it will get prompted earlier. 

Basically the way this works is that all the handlers are called in priority order until one handles it. 
I was just recommended to use priority 2000 because we don't have that many commands that we are handling, so it's better if we are further down the list.

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+        /// <summary>
+        /// Whether or not this handler should be ignored when multiple nodes are selected.
+        /// </summary>
+        public bool IgnoreOnMultiselect => true;
+
+        public int Exec(List<WorkspaceVisualNodeBase> selection, Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, IntPtr pvaIn, IntPtr pvaOut)
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:40]: Is null check for the "selection" needed?

Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:40]: If this method is invoked from outside of NuGet, does it need to be resilient to exceptions? Is it allowed as this method to throw?

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                            return 0;
+                    }
+                }
+            }
+            return 1;
+        }
+
+        public bool QueryStatus(List<WorkspaceVisualNodeBase> selection, Guid pguidCmdGroup, uint nCmdID, ref uint cmdf, ref string customTitle)
+        {
+            bool handled = false;
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:62]: Null check for the selection?

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+        private void ExecuteSolutionRestore(WorkspaceVisualNodeBase node)
+        {
+            // TODO: https://github.com/NuGet/Home/issues/9308
+        }
+
+        private static bool IsSolutionOnlySelection(List<WorkspaceVisualNodeBase> selection)
+        {
+            return selection.Count().Equals(1) && selection.First().NodeMoniker.Equals(string.Empty);
```
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs]: selection.Count() => selection.Count for better performance

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/OnlineEnvironment.cs (hunk 5 above)
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/OnlineEnvironment.cs:22]: Hopefully they do not change it. If they do we will catch in E2E testing right?

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs (hunk 8 above)
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:33]: Understand. Probably out of the picture now. But later would be any way to add telemetry to measure how fast they call into us?

**Review by cristinamanum — APPROVED** [review]
LGTM

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/OnlineEnvironment.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/OnlineEnvironment.cs:22]: Yeah, eventually we'll catch it in our E2E test as well. 
I'm working on test plans and infra on the side. 

These things are not supposed to be changed, specifically we are already doing things like this in our Current Visual Studio implementation. 

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.Tools/NuGetTools.vsct#L261

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs (hunk 8 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:33]: Given that this is on the UI thread, we have ongoing metrics to handle this scenario.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:33]: It's a VS level infra, not NuGet specific :)

File: scripts/cibuild/BuildValidator.ps1 (hunk 1 above)
Comment (nkolev92) [reply scripts/cibuild/BuildValidator.ps1]: Reverted and updated to the new version of the validator.

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs (hunk 9 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:40]: Think that's never null, but adjusted it to ensure NuGet does not throw because the docs don't explicitly say anything.

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs (hunk 10 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs:62]: Similar resolution to above.

File: src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs (hunk 11 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.OnlineEnvironment.Client/SolutionExplorer/NuGetWorkspaceCommandHandler.cs]: Nice catch. I missed the fact that it's a list.

**Comment by nkolev92** [comment]
Thanks for the review @cristinamanum, addressed all outstanding feedback.
